cmd/eval.go

```diff
@@ -83,9 +85,78 @@ func evalCmd() *cobra.Command {
 			return nil
 		},
 		RunE: func(cmd *cobra.Command, args []string) error {
-			//TODO: refactor once we handle application snapshots
-			if arguments.input != "" || arguments.filepath != "" {
-				report, err := applicationsnapshot.Report(snapshotSpec)
+			validateImage := func(imageRef string) (*policy.Output, error) {
+				out := &policy.Output{}
+
+				i, err := image.NewImageValidator(cmd.Context(), imageRef, arguments.publicKey, arguments.rekorURL)
+				if err != nil {
+					return nil, err
+				}
+
+				if err := i.ValidateImageSignature(cmd.Context()); err != nil {
+					out.SetImageSignatureCheck(false, err.Error())
+					return nil, err
+				}
+				out.SetImageSignatureCheck(true, "success")
+
+				if err := i.ValidateAttestationSignature(cmd.Context()); err != nil {
+					out.SetAttestationSignatureCheck(false, err.Error())
+					return nil, err
+				}
+				out.SetAttestationSignatureCheck(true, "success")
+
+				p, err := policy.NewPolicyEvaluator(arguments.policyConfiguration)
+				if err != nil {
+					return nil, err
+				}
+
+				results, err := p.Evaluate(cmd.Context(), i.Attestations())
+				if err != nil {
+					return nil, err
+				}
+				out.SetPolicyCheck(results)
+
+				return out, nil
+			}
+
+			if snapshotSpec != nil {
+
+				var wg sync.WaitGroup
+				component := make(chan applicationsnapshot.Component, 1)
+				cherr := make(chan error, 1)
+				var errs []error
+				var components []applicationsnapshot.Component
+
+				for _, c := range snapshotSpec.Components {
+					wg.Add(1)
+
+					go func(comp appstudioshared.ApplicationSnapshotComponent) {
+						defer wg.Done()
+
+						out, err := validateImage(comp.ContainerImage)
+						cherr <- err
+						errs = append(errs, <-cherr)
```

Comment:
This looks odd, shouldn't we fetch from the channel outside of the gorutine?

Reply:
Yes, moved it.